internal/transport/handler_server.go

```diff
@@ -264,29 +272,43 @@ func (ht *serverHandlerTransport) writeCommonHeaders(s *Stream) {
 	}
 }
 
+// writeCustomHeaders sets custom headers set on the stream via SetHeader
+// on the first write call (Write, WriteHeader, or WriteStatus).
+func (ht *serverHandlerTransport) writeCustomHeaders(s *Stream) {
+	h := ht.rw.Header()
+
+	s.hdrMu.Lock()
+	for k, vv := range s.header {
+		if isReservedHeader(k) {
+			continue
+		}
+		for _, v := range vv {
+			h.Add(k, encodeMetadataHeader(k, v))
+		}
+	}
+
+	s.hdrMu.Unlock()
+}
+
 func (ht *serverHandlerTransport) Write(s *Stream, hdr []byte, data []byte, opts *Options) error {
+	s.updateHeaderSent()
 	return ht.do(func() {
-		ht.writeCommonHeaders(s)
+		ht.writePendingHeaders(s)
 		ht.rw.Write(hdr)
 		ht.rw.Write(data)
 		ht.rw.(http.Flusher).Flush()
 	})
 }
 
 func (ht *serverHandlerTransport) WriteHeader(s *Stream, md metadata.MD) error {
+	if err := s.SetHeader(md); err != nil {
```

Comment:
tbh I don't really understand why `WriteHeader` even takes an `md` parameter - IMHO it would be more elegant to implement `(*Stream).SendHeader` via
```
s.SetHeader(md)
s.st.WriteHeader(s)
```
Since `ServerTransport` is an internal interface, I don't see any issues, but it would require changing the `http2Server` implementation as well, which is why I opted for a more local change.